core/src/consensus.rs

```diff
@@ -73,14 +83,32 @@ pub(crate) struct ComputedBankState {
 }
 
 #[frozen_abi(digest = "2ZUeCLMVQxmHYbeqMH7M97ifVSKoVErGvRHzyxcQRjgU")]
-#[derive(Serialize, AbiExample)]
+#[derive(Clone, Serialize, Deserialize, Debug, PartialEq, AbiExample)]
 pub struct Tower {
     node_pubkey: Pubkey,
     threshold_depth: usize,
     threshold_size: f64,
     lockouts: VoteState,
     last_vote: Vote,
     last_timestamp: BlockTimestamp,
+    #[serde(skip)]
+    path: PathBuf,
+    #[serde(skip)]
+    tmp_path: PathBuf, // used before atomic fs::rename()
+    #[serde(skip)]
+    // Restored voted slots which cannot be found in SlotHistory at replayed root
+    // (This's kind of special field for slashing-free validator restart with edge cases).
+    // Those voted slots means they are/were on a different fork (which may not be available
+    // in bank_forks after validator restart) or are unrooted slots following the replayed
+    // rooted slot.
+    // This is only used to construct synthesized ancestors for the last vote, if restored
+    // last_vote's ancestors cannot be found in bank_forks. That's because fork choice
+    // and switch threshold requires that information.
+    // That way we avoid mangling bank_forks or relaxing bank_forks invariants just for
+    // validator's restart porpose.
+    // This could be emptied after some time; but left intact indefinitely for easier
+    // implementation
+    stray_restored_slots: BTreeSet<Slot>,
```

Comment:
a new adjective `stray` is introduced to mean slots which aren't found in `bank_forks`.

Reply:
Being `BTreeSet` not `HashSet` is needed because of https://github.com/solana-labs/solana/pull/10718/files#diff-d848cba681535238849e7e33ed32702bR716.

Reply:
github isn't showing me the diff in the link 😢